py-polars/tests/unit/operations/aggregation/test_horizontal.py

```diff
@@ -556,3 +556,68 @@ def test_horizontal_sum_boolean_with_null() -> None:
     )
 
     assert_frame_equal(out.collect(), expected_df)
+
+
+@pytest.mark.parametrize("ignore_nulls", [True, False])
+@pytest.mark.parametrize(
+    ("dtype_in", "dtype_out"),
+    [
+        (pl.Null, pl.Null),
+        (pl.Boolean, pl.UInt32),
+        (pl.UInt8, pl.UInt8),
+        (pl.Float32, pl.Float32),
+        (pl.Float64, pl.Float64),
+        (pl.Decimal(None, 5), pl.Decimal(None, 5)),
+    ],
+)
+def test_horizontal_sum_with_null_col_ignore_strategy(
+    dtype_in: PolarsDataType,
+    dtype_out: PolarsDataType,
+    ignore_nulls: bool,
+) -> None:
+    lf = pl.LazyFrame(
+        {
+            "null": [None, None, None],
+            "s": pl.Series([1, 0, 1], dtype=dtype_in, strict=False),
+            "s2": pl.Series([1, 0, None], dtype=dtype_in, strict=False),
+        }
+    )
+    result = lf.select(pl.sum_horizontal("null", "s", "s2", ignore_nulls=ignore_nulls))
+    if ignore_nulls and dtype_in != pl.Null:
+        values = [2, 0, 1]
+    else:
+        values = [None, None, None]  # type: ignore[list-item]
```

Comment:
This is wrong. If our sum doesn't ignore nulls, it doesn't propagate them, but replaces them with the identity: 0.

The horizontal semantics should be the same as the vertical semantics.

Reply:
I think you have it reversed.

* Ignore nulls: 1 + 2 + null = 3 (nulls replaced by 0)
* Don't ignore nulls: 1 + 2 + null = null (nulls are propagated)

Reply:
I mean that our `sum` is agnostic to nulls. I think we made a mistake exposing this to `sum_horizontal` as our vertical sum is agnostic to nulls.

Reply:
I think what you're saying is that the `ignore_nulls` parameter should be removed entirely, and the `ignore_nulls=True` behavior should simply be the default. Is that correct? If so, that sounds like a breaking change.

Given that the current implementation has the parameter, and that there is an issue with it (`pl.Null` columns are not subject to the parameter, but nulls in other columns are), is the path forward to accept this PR (if I didn't mess anything up! which I don't think I did), and 2) remove the parameter in a follow-up PR to be merged in 1.19.0?

This PR contains a small fix for the float32 case where `mean_horizontal` returns f64 for f32 columns. I could make that a separate PR as well if you don't want to accept this one.

Reply:
Yeah, I think you're right. Consider it an observation. ;) Will take a look a bit later.

Reply:
Thanks Ritchie. I have a follow-up to this adding temporals for mean horizontal but I'll wait for this one first.